### ISSUE TITLE
System.InvalidOperationException: Error deserializing member

### ISSUE THREAD

**Issue by Anhaltiner0708** [jefffhaynes/BinarySerializer#50]
Hi Jeff,

I have updated the BinarySerializer from version 5.0.3 to the latest version. Since that I get an System.InvalidOperationException. Since that I get a System.InvalidOperationException. What is the reason and what should I do so that my code works again.

Cheers
Dirk

> System.OverflowException: Der Wert für einen Int32 war zu groß oder zu klein.
   bei System.Convert.ToInt32(UInt32 value)
   bei System.UInt32.System.IConvertible.ToInt32(IFormatProvider provider)
   bei System.Convert.ToInt32(Object value)
   bei BinarySerialization.TypeExtensions.ConvertTo(Object value, Type type)
   bei BinarySerialization.Graph.ValueGraph.ValueValueNode.ConvertToFieldType(Object value)
   bei BinarySerialization.Graph.ValueGraph.ValueValueNode.Deserialize(BinaryReader reader, SerializedType serializedType, Nullable`1 length)
   bei BinarySerialization.Graph.ValueGraph.ValueValueNode.Deserialize(BoundedStream stream, SerializedType serializedType, Nullable`1 length)
   bei BinarySerialization.Graph.ValueGraph.EnumValueNode.DeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ValueNode.DeserializeInternal(BoundedStream stream, Func`1 maxLengthDelegate, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
   --- Ende der internen Ausnahmestapelüberwachung ---
   bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ObjectValueNode.ObjectDeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ObjectValueNode.DeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ValueNode.DeserializeInternal(BoundedStream stream, Func`1 maxLengthDelegate, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
   --- Ende der internen Ausnahmestapelüberwachung ---
   bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.RootValueNode.DeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ValueNode.DeserializeInternal(BoundedStream stream, Func`1 maxLengthDelegate, EventShuttle eventShuttle)
   bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
   --- Ende der internen Ausnahmestapelüberwachung ---
   bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
   bei BinarySerialization.BinarySerializer.Deserialize(Stream stream, Type type, Object context)
   bei BinarySerialization.BinarySerializer.Deserialize(Byte[] data, Type type, Object context)

**Comment by jefffhaynes** [comment]
Sorry about that Dirk, I'll have to take a look tonight but it looks like it's failing when it tries to convert a binding source field to an int. 



		_____________________________

**Comment by Anhaltiner0708** [comment]
I have defined an enum without any restrictions. I mean the underlying type is integer. In my class I have assigned the following annotation.

`        [SerializeAs(SerializedType = SerializedType.UInt4)]
`

**Comment by jefffhaynes** [comment]
That does seem like it should work






On Wed, Sep 28, 2016 at 8:59 AM -0400, "Anhaltiner0708" <notifications@github.com> wrote:












I have defined an enum without any restrictions. I mean the underlying type is integer. In my class I have assigned the following annotation.



[SerializeAs(SerializedType = SerializedType.UInt4)]

**Comment by Anhaltiner0708** [comment]
It seems that you have added some stronger rules for the validation of the data type. At Version 5.0.3 this annotation (e.g SerializedType.UInt4) was working with the enum and the underlying type integer. But now the enum must be of the same type (e.g. uint). 

You have to define a enum for each different output, although the enumeration is identical. Is there a way, that the enum with the underlying type integer works for each annotation?

**Comment by jefffhaynes** [comment]
Yes, I think you're right. Let me see if it's necessary to make it so strict.






On Thu, Sep 29, 2016 at 12:31 AM -0400, "Anhaltiner0708" <notifications@github.com> wrote:












It seems that you have added some stronger rules for the validation of the data type. At Version 5.0.3 this annotation (e.g SerializedType.UInt4) was working with the enum and the underlying type integer. But now the enum must be of the same type (e.g. uint). 



You have to define a enum for each different output, although the enumeration is identical. Is there a way, that the enum with the underlying type integer works for each annotation?

**Comment by Anhaltiner0708** [comment]
Hey Jeff,

I have adapted the enum so that the underlying type is uint. This works well so far. However, when deserialize with big endianess following exception occurs. This is happen with the value 142/0x8E.

> --InvalidOperationException
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.RootValueNode.DeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.DeserializeInternal(BoundedStream stream, Func`1 maxLengthDelegate, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
> --InvalidOperationException
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ObjectValueNode.ObjectDeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ObjectValueNode.DeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.DeserializeInternal(BoundedStream stream, Func`1 maxLengthDelegate, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
> --OverflowException
>    bei System.Convert.ToInt32(UInt32 value)
>    bei System.UInt32.System.IConvertible.ToInt32(IFormatProvider provider)
>    bei System.Convert.ToInt32(Object value)
>    bei BinarySerialization.TypeExtensions.ConvertTo(Object value, Type type)
>    bei BinarySerialization.Graph.ValueGraph.ValueValueNode.ConvertToFieldType(Object value)
>    bei BinarySerialization.Graph.ValueGraph.ValueValueNode.Deserialize(BinaryReader reader, SerializedType serializedType, Nullable`1 length)
>    bei BinarySerialization.Graph.ValueGraph.EnumValueNode.DeserializeOverride(BoundedStream stream, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.DeserializeInternal(BoundedStream stream, Func`1 maxLengthDelegate, EventShuttle eventShuttle)
>    bei BinarySerialization.Graph.ValueGraph.ValueNode.Deserialize(BoundedStream stream, EventShuttle eventShuttle)
> Ergebnis Meldung:	
> System.InvalidOperationException : Error deserializing 'type 'Mhs.Gateway.Conveyor.Eag.Messages.Common.MsgHeader''.  See inner exception for detail.
>   ----> System.InvalidOperationException : Error deserializing 'type 'Mhs.Gateway.Conveyor.Eag.Messages.Common.MsgHeader''.  See inner exception for detail.
>   ----> System.InvalidOperationException : Error deserializing 'member 'PayloadType''.  See inner exception for detail.
>   ----> System.OverflowException : Der Wert für einen Int32 war zu groß oder zu klein.

**Comment by jefffhaynes** [comment]
Hey, I'm sorry I haven't had time to look at this but my laptop actually died last week. I'm getting a new one today and I'll really try to get to this as soon as possible.
Is you class definition somewhere on here already?



		_____________________________

**Comment by Anhaltiner0708** [comment]
Hi Jeff,

here is my class definition.

Thank you
Dirk

```C#
public enum PlcMessageType : ushort
	{
		Nok = 0, 
        
		Ack = 1, 
        
		Alive = 2, 
        
		Data = 3, 
        
		AckAlive = 4
	}

    public enum PlcPayloadType : uint{
        
        Undefined = 0,
        
        Value1 = 130,
        
        Value2 = 131,
        
        Value3 = 142,

        Value4 = 306,

        Value5 = 307
    }

public class MsgHeader {
        
        public MsgHeader()
            : base() {
        }
        
        [FieldOrder(0)]
        [SerializeAs(SerializedType = SerializedType.UInt2)]
        public PlcMessageType MsgType { get; set; }
        
        [FieldOrder(1)]
        [SerializeAs(SerializedType = SerializedType.UInt4)]
        public PlcPayloadType PayloadType { get; set; }
    }
```

**Comment by jefffhaynes** [comment]
Sorry for the delay, this is fixed in 5.8.1.  Thanks!
